### ISSUE TITLE
Problems preventing implementation of prehensile penis self-penetration

### ISSUE THREAD

**Issue by TadUnlikely** [Innoxia/liliths-throne-public#1078]
I've started some work on implementation of the PlayerSelfPenis and PartnerSelfPenis classes to allow for characters with the "prehensile" penis modifier to penetrate themselves in various ways. This Issue is for tracking what needs to be fixed before this can be sent as a pull request.

The first bug I'm seeing is that the SexSlotBipeds text entries are firing for self-penetration when they shouldn't be. In my test, my character Whoopsie stayed standing as the dominant partner. At the time of her orgasm, she chose to creampie herself. The orgasm description (essentially) started off with "You reach around and grab your ass, pulling you into you and letting out a moan as you prepare to reach your climax." This is awkward, and I'll need to figure out how to filter self-penetrative actions so that they either use more appropriate phrasing for these entries, or just skip them entirely.

The second bug is that it's not correctly calculating who got cum where. The remainder of the orgasm description looks correct, but if Whoopsie's "partner" has a penis, I get told that both of them cum inside her, and there's a chance they both got her pregnant. I'm not sure yet where that's being calculated or filtered, but it may be easier to start by fixing this.

Other than this, there's the question of flexibility and reachability. It seems logical that a cat-girl or a fox-girl would be more likely to be flexible enough for certain self-penetrative actions than a horse-boy or a human, for example, and the idea that somebody with a 1-inch member could effectively engage in any sort of self-penetrative action is pretty ludicrous. An 8-inch member is also going to be easier to self-suck for a 4-foot tall character than an 8-foot tall character.

The easier solution, and probably the one I'll start with, is just to do a flat filtering on penis length for now.  I can use the existing size categories to map and filter possibilities for different forms of prehensile self-penetration for the different orifices, and that will allow for implementing such actions earlier once the previously mentioned blocking bugs are resolved.

The more difficult solution, which I would like to work towards, is to implement reachability calculations that take into account penis length, height of character, and flexibility of the species. We'd probably also want to rebalance the size calculations for self-penetrative actions - I would rule that you "lose" effective inches due to having to bend around to penetrate yourself in certain way, and while I have intentions on allowing the knot modifier to be used for things like your mouth or nipple, it doesn't really make sense to be able to knot your own vagina or anus, unless we introduce a species with a knot mid-length rather than at the base. Right now that's just a descriptive difference, but if the game were to introduce the idea of continuing to allow for sex actions while waiting for your knot to deflate, that would become a mechanics question.

**Comment by CognitiveMist** [comment]
I'm intrigued. However, one thing I would like to point out is that the game doesn't currently do any reachability calculations for the other acts. For example, a one-inch penis wouldn't be very effective at paizuri-into-mouth, or anal knotting, or deepthroat, but you can still do all those.

If you do end up attempting such calculations, I think they would seem out of place if the existing actions were not also adjusted.

**Comment by TadUnlikely** [comment]
Absolutely. That would be a pretty serious overhaul, which is why I want to delay it until I can do it justice.

**Comment by jetharius** [comment]
I'm not sure if the framework for this has been started to expand the isSelfImpregnation bool in the litter class as I've not seen it referenced anywhere else outside of the phone stats where its set to return zero.
